### PR TITLE
Sync UUIDs of tables on replicas after RESTORE

### LINKED ISSUE #42709: Can't RESTORE engine=ReplicatedMergeTree('/path/{uuid}','{replica}') after successful BACKUP
**Describe what's wrong**
After successfull `BACKUP` for table `engine=ReplicatedMergeTree('/clickhouse/tables/{uuid}','{replica}')`
RESTORE TABLE 
**Does it reproduce on recent release?**
Yes

**How to reproduce**

**/etc/clickhouse-server/config.d/backup_storage_configuration.xml**

```xml
<?xml version="1.0"?>
<clickhouse>
    <backups>
        <allowed_path>/var/lib/clickhouse/backups_embedded/</allowed_path>
    </backups>
</clickhouse>
```

```sql
DROP DATABASE IF EXISTS t1 SYNC;
CREATE DATABASE t1 ENGINE=Atomic;
SELECT * FROM system.clusters FORMAT Vertical;
CREATE TABLE IF NOT EXISTS t1.r1 ON CLUSTER 'cluster' (v UInt64) ENGINE=ReplicatedMergeTree('/clickhouse/tables/{uuid}','{replica}') ORDER BY tuple();
INSERT INTO t1.r1 SELECT number FROM numbers(100);
BACKUP TABLE `t1`.`r1`TO File('test_uuid_replicated.zip');
DROP DATABASE IF EXISTS t1 SYNC;
CREATE DATABASE t1 ENGINE=Atomic;
RESTORE ALL FROM File('test_uuid_replicated.zip');
SELECT * FROM t1.r1;
```

* Which ClickHouse server version to use
reproduced on 22.10, 22.9, 22.8

**Expected behavior**
restore table 

**Error message and/or stacktrace**
```
RESTORE ALL FROM File('test_uuid_replicated.zip')

Received exception from server (version 22.10.1):
Code: 62. DB::Exception: Received from localhost:9000. DB::Exception: Macro 'uuid' and empty arguments of ReplicatedMergeTree are supported only for ON CLUSTER queries with Atomic database engine: While creating table t1.r1. (SYNTAX_ERROR)
```
**Additional context**

Full stack trace
```
2022.10.26 16:57:26.989648 [ 275 ] {97ed60e6-89e1-41b3-b1c4-8518c97e94eb} <Debug> executeQuery: (from 127.0.0.1:50638) RESTORE ALL FROM File('test_uuid_replicated.zip'); (stage: Complete)
2022.10.26 16:57:26.990557 [ 275 ] {97ed60e6-89e1-41b3-b1c4-8518c97e94eb} <Trace> RestorerFromBackup: Will use paths in backup: "/"
2022.10.26 16:57:26.990645 [ 275 ] {97ed60e6-89e1-41b3-b1c4-8518c97e94eb} <Trace> RestorerFromBackup: Finding tables in backup
2022.10.26 16:57:26.990944 [ 275 ] {97ed60e6-89e1-41b3-b1c4-8518c97e94eb} <Information> RestorerFromBackup: Will restore 0 databases and 1 tables
2022.10.26 16:57:26.991039 [ 275 ] {97ed60e6-89e1-41b3-b1c4-8518c97e94eb} <Trace> ContextAccess (default): Access granted: INSERT, CREATE TABLE ON t1.r1
2022.10.26 16:57:26.991061 [ 275 ] {97ed60e6-89e1-41b3-b1c4-8518c97e94eb} <Trace> RestorerFromBackup: Creating databases
2022.10.26 16:57:26.991077 [ 275 ] {97ed60e6-89e1-41b3-b1c4-8518c97e94eb} <Trace> RestorerFromBackup: Creating tables
2022.10.26 16:57:26.991166 [ 275 ] {97ed60e6-89e1-41b3-b1c4-8518c97e94eb} <Trace> RestorerFromBackup: Creating table t1.r1: CREATE TABLE IF NOT EXISTS t1.r1 (`v` UInt64) ENGINE = ReplicatedMergeTree('/clickhouse/tables/{uuid}', '{replica}') ORDER BY tuple() SETTINGS index_granularity = 8192
2022.10.26 16:57:26.991983 [ 275 ] {97ed60e6-89e1-41b3-b1c4-8518c97e94eb} <Error> executeQuery: Code: 62. DB::Exception: Macro 'uuid' and empty arguments of ReplicatedMergeTree are supported only for ON CLUSTER queries with Atomic database engine: While creating table t1.r1. (SYNTAX_ERROR) (version 22.10.1.1877 (official build)) (from 127.0.0.1:50638) (in query: RESTORE ALL FROM File('test_uuid_replicated.zip');), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xce3f35a in /usr/bin/clickhouse
1. DB::Macros::expand(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Macros::MacroExpansionInfo&) const @ 0xce4802f in /usr/bin/clickhouse
2. ? @ 0x1336acb6 in /usr/bin/clickhouse
3. DB::StorageFactory::get(DB::ASTCreateQuery const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, bool) const @ 0x12c4ac7b in /usr/bin/clickhouse
4. DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&, std::__1::unique_ptr<DB::DDLGuard, std::__1::default_delete<DB::DDLGuard> >&) @ 0x1240eea6 in /usr/bin/clickhouse
5. DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) @ 0x12409448 in /usr/bin/clickhouse
6. DB::InterpreterCreateQuery::execute() @ 0x12412d0d in /usr/bin/clickhouse
7. DB::DatabaseWithOwnTablesBase::createTableRestoredFromBackup(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>, std::__1::shared_ptr<DB::IRestoreCoordination>, unsigned long) @ 0x11a74b03 in /usr/bin/clickhouse
8. DB::RestorerFromBackup::createTable(DB::QualifiedTableName const&) @ 0x116865fe in /usr/bin/clickhouse
9. DB::RestorerFromBackup::createTables() @ 0x1166f8bb in /usr/bin/clickhouse
10. DB::RestorerFromBackup::run(DB::RestorerFromBackup::Mode) @ 0x1166a2c4 in /usr/bin/clickhouse
11. DB::BackupsWorker::doRestore(std::__1::shared_ptr<DB::ASTBackupQuery> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::RestoreSettings, DB::BackupInfo const&, std::__1::shared_ptr<DB::IRestoreCoordination>, std::__1::shared_ptr<DB::Context>, bool) @ 0x1163526a in /usr/bin/clickhouse
12. DB::BackupsWorker::startRestoring(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>) @ 0x1162e2ef in /usr/bin/clickhouse
13. DB::BackupsWorker::start(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>) @ 0x1162c992 in /usr/bin/clickhouse
14. DB::InterpreterBackupQuery::execute() @ 0x125b1513 in /usr/bin/clickhouse
```

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Improvement

### Changelog entry:
`RESTORE TABLE ON CLUSTER` must create replicated tables with a matching UUID on hosts. Otherwise the macro `{uuid}` in ZooKeeper path can't work correctly after RESTORE. This PR implements that.

This PR fixes https://github.com/ClickHouse/ClickHouse/issues/42709.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit a2895327c73599bafd56d5808094d7f866d68788 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53765/a2895327c73599bafd56d5808094d7f866d68788/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 1]
```diff
@@ -219,10 +219,12 @@ BlockIO InterpreterCreateQuery::createDatabase(ASTCreateQuery & create)
     else
     {
         bool is_on_cluster = getContext()->getClientInfo().query_kind == ClientInfo::QueryKind::SECONDARY_QUERY;
-        if (create.uuid != UUIDHelpers::Nil && !is_on_cluster)
+        if (create.uuid != UUIDHelpers::Nil && !is_on_cluster && !internal)
```
Comment (SmitaRKulkarni) [inline src/Interpreters/InterpreterCreateQuery.cpp:222]: Isn't `is_on_cluster` same as `internal`?

File: src/Backups/RestoreCoordinationLocal.cpp [hunk 2]
```diff
@@ -51,6 +52,40 @@ bool RestoreCoordinationLocal::acquireReplicatedSQLObjects(const String &, UserD
...
+    create_query.setUUID({});
+    auto new_uuids = create_query.generateRandomUUID();
+    String new_query_str = serializeAST(create_query);
+
+    {
+        std::lock_guard lock{mutex};
+        if (find_in_map())
+            return;
+        create_query_uuids[query_str] = new_uuids;
+        create_query_uuids[new_query_str] = new_uuids;
```
Comment (SmitaRKulkarni) [inline src/Backups/RestoreCoordinationLocal.cpp:85]: Why do we need to store both? we can store only the serialized AST, and always use serialized AST during checking right ?. maybe I am missing something?

File: src/Interpreters/InterpreterCreateQuery.cpp (hunk 1 above)
Comment (vitlibar) [reply src/Interpreters/InterpreterCreateQuery.cpp:222]: Not always. `RestorerFromBackup` calls [InterpreterCreateQuery::execute()](https://github.com/ClickHouse/ClickHouse/blob/886a6e6e86f750c2d53d1f2dbdddf76b74061d86/src/Databases/DatabasesCommon.cpp#L368) with `internal` set to true. It's not ON CLUSTER queries.

And `is_on_cluster` is set when `CREATE TABLE ON CLUSTER` is executed, it has nothing to do with backups.

Comment (vitlibar) [reply src/Interpreters/InterpreterCreateQuery.cpp:222]: I wrote some comment in the code.

File: src/Backups/RestoreCoordinationRemote.cpp [hunk 3]
```diff
@@ -231,6 +234,38 @@ bool RestoreCoordinationRemote::acquireReplicatedSQLObjects(const String & loade
...
+        [&, &zk = holder.faulty_zookeeper]()
+        {
+            with_retries.renewZooKeeper(zk);
+
+            String path = zookeeper_path + "/table_uuids/" + escapeForFileName(query_str);
+
+            Coordination::Requests ops;
+            ops.emplace_back(zkutil::makeCreateRequest(path, new_uuids_str, zkutil::CreateMode::Persistent));
+
+            Coordination::Responses responses;
```
Comment (SmitaRKulkarni) [inline src/Backups/RestoreCoordinationRemote.cpp:253]: unused?

**Review by SmitaRKulkarni — COMMENTED** [review]
Rest all LGTM

File: src/Backups/RestoreCoordinationRemote.cpp (hunk 3 above)
Comment (vitlibar) [reply src/Backups/RestoreCoordinationRemote.cpp]: yes, I've removed that

**Review by SmitaRKulkarni — APPROVED** [review]

**Comment by Alex-Cheng** [comment]
Could I "RESTORE" a ReplicatedMergeTree table in a non-cluster ClickHouse server? The scenario is that we may need to migrate data from product cluster environment to local development envrionment which is not cluster.

Of course, creating a local cluster enviornment for development is a workaround.

**Pull request merged**